### PR TITLE
KBC-2570 Edit test for new behavior of delete trigger

### PR DESCRIPTION
**Pull request by fejjo**
Jira: KBC-2570
Main PR: https://github.com/keboola/connection/pull/3689

Before asking for review make sure that:

- [x] New client method(s) has tests
- [x] Apiary file is updated
- [x] You declared if there is a BC break or not (will affect next release of a client)

---

### CODE REVIEW

File: tests/Common/TriggersTest.php [hunk 1]
```diff
@@ -542,34 +533,75 @@ public function testDeleteTriggerWithNonMasterToken()
...
+
+        $optionsForTokenRunWith = (new TokenCreateOptions());
+        $newNonAdminToken = $this->tokens->createToken($optionsForTokenRunWith);
+        $anotherClientWithoutPrivilage = $this->getClient(['url' => STORAGE_API_URL, 'token' => $newNonAdminToken['token']]);
+        try {
+            $anotherClientWithoutPrivilage->deleteTrigger((int) $trigger['id']);
+            self::fail('should fail');
+        } catch (Exception $e) {
+            self::assertEquals('Your token does not have sufficient privilege.', $e->getMessage());
+        }
```
Comment (tomasfejfar) [inline tests/Common/TriggersTest.php:604]: Tady by naopak mohla být expect exception snadno :)

**Review by tomasfejfar — CHANGES_REQUESTED** [review]

File: tests/Common/TriggersTest.php [hunk 2]
```diff
@@ -542,34 +533,73 @@ public function testDeleteTriggerWithNonMasterToken()
...
-        try {
-            $anotherClientWithAnotherToken->deleteTrigger((int) $trigger['id']);
-            self::fail('should fail');
-        } catch (Exception $e) {
-            self::assertEquals('Insufficient privilege. This token cannot manipulate with this trigger', $e->getMessage());
-        }
+        $anotherClientWithAnotherToken->deleteTrigger((int) $trigger['id']);
+        $this->expectException(ClientException::class);
+        $this->expectExceptionMessage(sprintf('Trigger with id [%d] was not found', $trigger['id']));
+        $clientWithoutAdminToken->getTrigger((int) $trigger['id']);
```
[the remark this answers is not in the document]
Comment (fejjo) [reply tests/Common/TriggersTest.php:553]: > Zbytek teda reviewnu, až tohle bude fixlé a bude jasné, že ty testy testují všechny edge cases. 

Jo jasne sry

File: tests/Common/TriggersTest.php [hunk 3]
```diff
@@ -542,34 +533,75 @@ public function testDeleteTriggerWithNonMasterToken()
...
+
+        $this->assertSame(false, true);
```
[the remark this answers is not in the document]
Comment (fejjo) [reply tests/Common/TriggersTest.php]: ty jo dobry fail no :/

**Comment by fejjo** [comment]
Testujem:
Trigger vytvoreny adminTokenom:
- nejde updatnut tokenom ktory nema opravnenia (canManageBuckets:false, opravnenie na orchestraciu)
- ide updatnut s tokenom s opravnenim 
- ide updatnut s master tokenom

Trigger vytvoreny nonAdmin tokenom
- ide updatnut nonAdmin tokenom s opravnenim ktory tento trigger vytvoril
- nejde updatnut s tokenom ktory nema opravnenia (canManageBuckets:false, opravnenie na orchestraciu)
- ide updatnut adminTokenom

Trigger vytvoreny nonAdmin tokenom:
- ide updatnut nonAdmin tokenom ktory ma opravnenia a nezalozil dani trigger
-

**Comment by fejjo** [comment]
Testujem 

Trigger vytvoreny master tokenom
- nejde zmazat nonAdmin tokenom bez permission (canManageBuckets:false, opravnenie na orchestaciu)
- ide zmazat nonAdmin tokenom s opravnenimi
- ide zmazat admin tokenom

Trigger vytvoreny NonMastertokenom
- nejde zmazat nonAdmin tokenom bez opravneni
- ide zmazat non admin tokenom s opravnenym ktory ho nevytvoril
- ide zmazat admin tokenom
Trigger vytvorenu nonMaster tokenom
- ide zmazat tokenom ktory trigger nevytvoril a ma spravne privileges

File: tests/Common/TriggersTest.php [hunk 4]
```diff
@@ -492,31 +493,51 @@ public function testDeleteTriggerWithMasterToken()
...
+            'runWithTokenId' => $newToken['id'],
+            'tableIds' => [
+                $table,
+            ],
+        ]);
 
-        $newNonAdminToken = $this->tokens->createToken((new TokenCreateOptions())->setCanManageBuckets(true));
-        $clientWithoutAdminToken = $this->getClient(['url' => STORAGE_API_URL, 'token' => $newNonAdminToken['token']]);
+        $newNonAdminTokenWithoutPermission = $this->tokens->createToken((new TokenCreateOptions()));
+        $clientWithoutAdminTokenWithoutPermission = $this->getClient(['url' => STORAGE_API_URL, 'token' => $newNonAdminTokenWithoutPermission['token']]);
```
Comment (tomasfejfar) [inline tests/Common/TriggersTest.php:507]: Tohle je dost divný pattern v kombinaci s tím data providerem - vlastně to otestuješ 4x pro každý z těch dataproviderů :/ 

Ale už to nechme. 

Ideálně by to bylo:

`testTriggerCreatedByMasterTokenCanBeDeletedByToken` (tady by byly ty 4 tokeny co to s nima jde)
`testTriggerCreatedByMasterTokenCannotBeDeletedByToken` (tady by byl ten jeden co to s ním nejde)

File: tests/Common/TriggersTest.php [hunk 5]
```diff
@@ -543,33 +564,77 @@ public function testDeleteTriggerWithNonMasterToken()
...
+    /**
+     * @param int $triggerId
+     * @param array $triggerList
+     * @return void
+     */
+    private function assertNotContainsTriggerId($triggerId, $triggerList)
+    {
+        $this->assertFalse(in_array($triggerId, array_map(function ($trigger) {
+            return $trigger['id'];
+        }, $triggerList)));
```
Comment (tomasfejfar) [inline tests/Common/TriggersTest.php:637]: AssertFalse má nevýhodu, že ti neřekne nic o tom, co bylo v tom poli. Je lepší assertovat že "není v poli"...

Comment (tomasfejfar) [reply tests/Common/TriggersTest.php:637]:
```
Failed asserting that true is false.
 /code/tests/Common/TriggersTest.php:637
 /code/tests/Common/TriggersTest.php:526
 /code/vendor/bshaffer/phpunit-retry-annotations/src/RetryTrait.php:33

```

Comment (tomasfejfar) [reply tests/Common/TriggersTest.php:637]:
```suggestion
        $this->assertNotContains($triggerId, array_map(function ($trigger) {
            return $trigger['id'];
        }, $triggerList));
```

Ale koukám, že to taky nevrací co v tom poli teda bylo :/

Comment (tomasfejfar) [reply tests/Common/TriggersTest.php:637]: Ale furt mi to přijde jako čitelnější error :)

File: tests/Common/TriggersTest.php [hunk 6]
```diff
@@ -543,33 +564,77 @@ public function testDeleteTriggerWithNonMasterToken()
...
-        $this->expectExceptionMessage(sprintf('Trigger with id [%d] was not found', $trigger2['id']));
-        $this->_client->getTrigger((int) $trigger2['id']);
+        $this->assertCount(0, $triggerList);
+    }
+
+    /**
+     * @dataProvider tokenCreateOptionsProvider
+     * @return void
+     */
+    public function testDeleteTriggerCreatedByNonMasterToken(TokenCreateOptions $optionForToken)
```
Comment (tomasfejfar) [inline tests/Common/TriggersTest.php:602]: Tohle nechápu, co testuje. To testuje, že když to může vytvořit, tak to může i smazat? To mi skoro přijde nadbytečné. 

Možná to aspoň přejmenovat
```suggestion
    public function testNonMasterTokenCanDeleteTriggerItCreated(TokenCreateOptions $optionForToken)
```

File: tests/Common/TriggersTest.php [hunk 7]
```diff
@@ -492,31 +493,51 @@ public function testDeleteTriggerWithMasterToken()
...
+
+        $this->assertCount(0, $triggerList);
     }
 
     /**
+     * @dataProvider tokenCreateOptionsProvider
      * @return void
      */
-    public function testDeleteTriggerWithNonMasterToken()
+    public function testDeleteTriggerCreatedByNonMasterTokenDifferentToken(TokenCreateOptions $optionForToken)
```
Comment (tomasfejfar) [inline tests/Common/TriggersTest.php:540]:
```suggestion
    public function testDeleteTriggerCreatedByNonMasterTokenUsingDifferentToken(TokenCreateOptions $optionForToken)
```

**Review by tomasfejfar — APPROVED** [review]
Jen kosmetické věci
